designs/Generic_Builder_Interface.md

```diff
@@ -0,0 +1,92 @@
+
+The logic for mapping a project's structure to a workflow (e.g. looking for `pom.xml` or `build.gradle` to determine the java project type) is currently bundled into SAM CLI instead of `aws-lambda-builders`, burdening and limiting consumers.
+
+Installing the `aws-lambda-builders` tool requires the use of `pip install`, which complicates the setup process of consumers such as the CDK (installed with `npm`) because instructions must be provided and followed by all customers to successfully set up their environment. Developers familiar with `npm` and looking to build a CDK project are now required to configure and manage tooling outside their initial intention. We must keep enviromental dependencies to a minimum.
+
+What will be changed?
+---------------------
+Extract the automatic project inference from `sam cli` into this library and extend the RPC interface so that projects other than Lambda functions may also be built and packaged. 
+
+Rename or create a new package, `aws-builders`, to better represent the (now) wider-surface area. 
+
+Publish Docker images bundling the CLI for various target language/platforms so that consumers need not depend on `pip install` to consume the CLI. Instead, the tool can be invoked with [`docker run`](https://docs.docker.com/engine/reference/commandline/run/), requiring only a single dependency on the Docker toolchain. 
+
+Taking a dependency on Docker is justified as follows:
+* `aws-lambda-builders` already leverages Docker to build native code compatible with Lambda
+* CDK already depends on Docker to build and push images to ECR, so no extra dependencies are required there.
+* Docker successfully encapsulates the complexity of OS platforms and tool installations, allowing us to scale to a wide variety of use-cases without burdening our customers with toolchain dependencies.
+
+Success criteria for the change
+-------------------------------
+A single library and CLI exists, encapsulating the logic to build deployable artifacts from arbitrary code workspaces. It should attempt to infer boiler-plate details such as maven vs gradle for building java, requiring only the language target from the caller. "This is a java 1.8 project, build it for me".
+
+This CLI may be installed directly into the developer's environment or consumable as encapsulated Docker images specific to various langauges/environments.
+
+User Experience Walkthrough
+---------------------------
+
+*TODO: More details/use-cases*
+
+For the CDK use-case, we wish to instantiate a language-specific function type and only have to point to the project's location:
+
+```ts
+const myFunction = new lambda.JavaFunction(stack, 'MyJavaFunction', {
```

Comment:
Couldn't we make use of the newly centralized build/language detection logic to get rid of the need of specifying `Java` here?

Reply:
Also, strictly speaking, shouldn't this be called `JvmFunction`? :)

Reply:
There's a different APIs for defining the build project for different languages. Different way to specify dependencies/handlers/runtime versions. I like the strong typing.